painless_mesh_1.ino

```diff
@@ -0,0 +1,282 @@
+// Needed for painless library
+void receivedCallback( uint32_t from, String &msg ) {
+  bool complete = true;
+  Serial.printf("startHere: Received from %u msg=%s\n", from, msg.c_str());
+  // state machine (state 1 = layer 1 ...)
+  buff = msg;
+  // 1st character in message denotes which message
+  // check if received, if received ignore, if not store to memory
+  if (nodeReceivedState[layer_state][hashMap.getValueOf(from)][buff[1]]== 0 ){ // if message not received => save to buffer
+    // save to buffer
+    Serial.println("save to buffer");
+  } 
+
+  // check whether this layer is done
+  for (int i=0 ; i<layered_numNodes[layer_state] ; ++i ) { // go through all nodes within layer
+    for(int j = 0 ; j <3 ; ++i ){
+      if (nodeReceivedState[layer_state][i][j]==0){// checks message state
+        complete = false; // so if we all 1, our complete flag stays true means we completed
+      } 
+    }
+  }
+  // if complete flag still true we go to next layer
+  if (complete == true){
+    layer_state += 1;  // move on to next layer
+  }
+}
+
+void newConnectionCallback(uint32_t nodeId) {
+  Serial.printf("--> startHere: New Connection, nodeId = %u\n", nodeId);
+  Serial.printf("--> startHere: New Connection, %s\n", mesh.subConnectionJson(true).c_str());
+  // prints mesh subconnections
+}
+
+void changedConnectionCallback() {
+  Serial.printf("Changed connections\n");
+  nodes = mesh.getNodeList();
+  numNodes = nodes.size();
+  Serial.printf("Num nodes: %d\n", numNodes);
+  Serial.printf("Connection list:");
+  // does not contain itself
+  SimpleList<uint32_t>::iterator node = nodes.begin();
+  int nodeCount = 0;
+  while (node != nodes.end()) 
+    {
+    nodeList[nodeCount] = *node ; // adds subnodes into list
+    //Serial.println(nodeList[nodeCount]);
+    Serial.printf(" %u", *node);
+    node++;
+    
+    nodeCount +=1;
+    }
+  
+  Serial.println();
+}
+
+void nodeTimeAdjustedCallback(int32_t offset) {
+  Serial.printf("Adjusted time %u. Offset = %d\n", mesh.getNodeTime(), offset);
+}
+
+// done portion needed by painless mesh
+
+void sendMessage(){
+
+  
+  int length = 6;
+  char msg[length]={0,}; // initialise x sized char => with 0's
+  for (int i = 0 ;i <length-1; i++  ){
+    msg[i] = 'a';
+    //Serial.println(i);
+  } 
+  msg[length -1] = '\0';
+  //msg = "This is a trial";
+  Serial.println(sizeof(msg));
+  Serial.println(msg[5]);
+  //serializeJson(doc, msg);
+  mesh.sendBroadcast( msg );
+  Serial.print("Mesh Broadcast - "); Serial.println(msg);
+  Serial.printf("--> startHere: New Connection, %s\n", mesh.subConnectionJson(true).c_str());
+  
+  
+  // parsing Json
+  
+
+  DeserializationError error = deserializeJson(doc, mesh.subConnectionJson(true).c_str());
+
+  if (error) {
+    Serial.print("deserializeJson() failed: ");
+    Serial.println(error.c_str());
+    return;
+  }
+
+  long long nodeId = doc["nodeId"]; // 2127231457
+  bool root = doc["root"]; // true
+
```

Comment:
A less repetitive solution should be attempted for mesh network structure decoding